### PR TITLE
Search: API V3

### PR DESCRIPTION
**Pull request by stsewd**
Implementation of https://github.com/readthedocs/readthedocs.org/pull/9533

Fixes https://github.com/readthedocs/readthedocs.org/issues/9573

This is on top of https://github.com/readthedocs/readthedocs.org/pull/9624

At first, I started to find a way to share most of the code with API v2, but the code resulted hard to follow,
so I just did a clean start for the new API (well, almost, there are still a lot of things that are being re-used).

- The new syntax can be used in the global search
- The project search now redirects to the global search with the current project set in the query.
- Some facets have been removed to just support the project/language facets.
- The global search now defaults to `type=file`.
- You can still search projects.
- The project facet isn't used directly, but only to aggregate results in the dashboard.
- The documentation was re-organized into a new directory instead of just one page.
- The advanced search guide was deleted, and their contents are now part of the documentation itself.

---
:books: Documentation previews :books:

- User's documentation (`docs`): https://docs--9625.org.readthedocs.build/en/9625/

- Developer's documentation (`dev`): https://dev--9625.org.readthedocs.build/en/9625/

### CODE REVIEW

**Comment by stsewd** [comment]
This should be ready, but we should first merge https://github.com/readthedocs/readthedocs.org/pull/9624, so I can fix the merge conflicts :/

File: readthedocs/search/api/v3/backend.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
+from functools import cached_property
+from itertools import islice
+
+from readthedocs.builds.models import Version
+from readthedocs.projects.models import Project
+from readthedocs.search.api.v3.queryparser import SearchQueryParser
+from readthedocs.search.faceted_search import PageSearch
+
+
+class Backend:
```
Comment (stsewd) [inline readthedocs/search/api/v3/backend.py:10]: I have no idea how to name this "thing", backend was the thing the came to my head, my I don't like it. This is just an abstraction that we can re-use in the api and the global search view.

File: readthedocs/search/api/v3/backend.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    """
+
+    def __init__(
+        self,
+        *,
+        request,
+        query,
+        arguments_required=True,
+        default_all=False,
+        max_projects=100
```
Comment (stsewd) [inline readthedocs/search/api/v3/backend.py:32]: We can decrease or increase this number if we want, I just chose a number that isn't that big, an org with lots of projects may need a larger number? I'll check the numbers on .org and .com...

File: readthedocs/search/api/v3/backend.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        self.arguments_required = arguments_required
+        self.default_all = default_all
+        self.max_projects = max_projects
+
+    @cached_property
+    def projects(self):
+        """
+        Return all projects used in this search.
+
+        If empty, it will search all projects.
```
Comment (stsewd) [inline readthedocs/search/api/v3/backend.py:45]: don't really like that empty means all, but not sure how to represent all without having to include ALL projects (this is mainly for .org where we allow searching ALL projects)

**Comment by stsewd** [comment]
I was also thinking to split the doc refactor in another PR if that makes things simpler (only the refactor, no adding new docs)

File: readthedocs/search/api/v3/urls.py [hunk 4]
```diff
@@ -0,0 +1,7 @@
+from django.urls import path
+
+from readthedocs.search.api.v3.views import SearchAPI
+
+urlpatterns = [
+    path("", SearchAPI.as_view(), name="search_api_v3"),
```
Comment (ericholscher) [inline readthedocs/search/api/v3/urls.py:6]: This feels like an odd pattern. Should we just include it directly?

File: readthedocs/search/api/v3/utils.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+from readthedocs.projects.models import Feature
+
+
+def should_use_advanced_query(projects):
+    # TODO: we should make this a parameter in the API,
+    # we are checking if the first project has this feature for now.
```
Comment (ericholscher) [inline readthedocs/search/api/v3/utils.py:6]: This is definitely pretty weird/janky. It would be best if we could avoid having this logic entirely, but I don't 100% know the best approach.

File: readthedocs/search/api/v3/backend.py (hunk 1 above)
Comment (ericholscher) [reply readthedocs/search/api/v3/backend.py:10]: `SearchExecuter` or `SearchWrapper` perhaps? I don't fully understand what it's doing, but we can definitely come up with a better name, and it should have `Search` in it as well.

File: readthedocs/search/api/v3/backend.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+            **kwargs,
+        )
+        return search
+
+    def _get_projects_to_search(self):
+        if not self._has_arguments:
+            if self.arguments_required:
+                return None
+            yield from self._get_default_projects()
+            return None
```
Comment (ericholscher) [inline readthedocs/search/api/v3/backend.py:85]: I'm confused by what this is trying to do? It's both yielding and returning?

File: readthedocs/search/api/v3/backend.py [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+
+        for value in self.parser.arguments["project"]:
```
Comment (ericholscher) [inline readthedocs/search/api/v3/backend.py:87]: `value` is not a great variable name, what is it, a project slug? We should update the variables below as well.

File: readthedocs/search/api/v3/backend.py [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+        for value in self.parser.arguments["subprojects"]:
+            project, version = self._get_project_and_version(value)
+
+            # Add the project itself.
+            if version and self._has_permission(self.request.user, version):
+                yield project, version
+
+            # If the user didn't provide a version, version_slug will be `None`,
+            # and we add all subprojects with their default version,
+            # otherwise we will add all projects that match the given version.
```
Comment (ericholscher) [inline readthedocs/search/api/v3/backend.py:101]: Only if they exist though, right?

File: readthedocs/search/api/v3/views.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+            request=self.request,
+            query=self.request.GET["q"],
+        )
+        return backend
+
+    def _get_search_query(self):
+        return self._backend.parser.query
+
+    def _get_projects_to_search(self):
+        return self._backend.projects
```
Comment (ericholscher) [inline readthedocs/search/api/v3/views.py:59]: I'm not sure if wrapping this makes it more readable. Probably easiest to just include them directly, unless we expect this logic to change?

File: readthedocs/search/api/v3/views.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+        for project, version in self._get_projects_to_search():
+            tasks.record_search_query.delay(
+                project.slug,
+                version.slug,
+                query,
+                total_results,
+                time.isoformat(),
+            )
+
+    def list(self):
```
Comment (ericholscher) [inline readthedocs/search/api/v3/views.py:104]: Would be useful to understand why we're overriding this method. A comment/docstring would be good.

File: readthedocs/search/api/v3/views.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+        )
+        serializer = self.get_serializer(
+            page, many=True, projects=self._get_projects_to_search()
+        )
+        response = self.paginator.get_paginated_response(serializer.data)
+        self._add_extra_fields(response)
+        return response
+
+    def _add_extra_fields(self, response):
+        """Add additional fields to the response."""
```
Comment (ericholscher) [inline readthedocs/search/api/v3/views.py:119]: This is not a good docstring :) Why are we adding these, why here?

File: readthedocs/search/api/v3/views.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+        response.data["projects"] = [
+            {"slug": project.slug, "versions": [{"slug": version.slug}]}
+            for project, version in self._get_projects_to_search()
+        ]
+        response.data["query"] = self._get_search_query()
+
+
+class BaseProxiedSearchAPI(SearchAPI):
```
Comment (ericholscher) [inline readthedocs/search/api/v3/views.py:127]: This should explain why it's here since it seems odd without an explanation -- I assume for some downstream reason?

File: readthedocs/templates/core/project_bar_base.html [hunk 13]
```diff
@@ -41,7 +41,7 @@ <h1>
 
         <li class="{{ downloads_active }}"><a href="{% url "project_downloads" project.slug %}" rel="nofollow,noindex">{% trans "Downloads" %}</a></li>
 
-        <li class="{{ search_active }}"><a href="{% url "elastic_project_search" project.slug %}" rel="nofollow,noindex">{% trans "Search" %}</a></li>
+        <li class="{{ search_active }}"><a href="{% url "search" %}?q=project:{{ project.slug|urlencode }}%20" rel="nofollow,noindex">{% trans "Search" %}</a></li>
```
Comment (ericholscher) [inline readthedocs/templates/core/project_bar_base.html:44]: This feels like a weird UX.. I think we probably just remove this from the menu until we can make a nicer way to make a global search bar work well? We can definitely have "search this project" in some other UI element though, but in the menu it feels odd to end up outside the project context.

File: readthedocs/urls.py [hunk 14]
```diff
@@ -90,7 +90,8 @@
 api_urls = [
     re_path(r'^api/v2/', include('readthedocs.api.v2.urls')),
     # Keep `search_api` at root level, so the test does not fail for other API
-    re_path(r"^api/v2/search/$", include("readthedocs.search.api.v2.urls")),
+    path("api/v2/search/", include("readthedocs.search.api.v2.urls")),
+    path("api/v3/search/", include("readthedocs.search.api.v3.urls")),
```
Comment (ericholscher) [inline readthedocs/urls.py:94]: We should probably keep using re_path for consistency?

File: readthedocs/search/api/v3/views.py [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+from readthedocs.search.api.v3.serializers import PageSearchSerializer
+from readthedocs.search.api.v3.utils import should_use_advanced_query
+
+log = structlog.get_logger(__name__)
+
+
+class SearchAPI(GenericAPIView):
+
+    """
+    Server side search API V3.
```
Comment (ericholscher) [inline readthedocs/search/api/v3/views.py:22]: Are we using the same auth & semantics as the rest of API v3? If not, this will be confusing to users, and hard to document.

File: readthedocs/search/api/v3/backend.py [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+            projects=projects,
+            **kwargs,
+        )
+        return search
+
+    def _get_projects_to_search(self):
+        if not self._has_arguments:
+            if self.arguments_required:
+                return None
+            yield from self._get_default_projects()
```
Comment (ericholscher) [inline readthedocs/search/api/v3/backend.py:84]: It feels odd that this code is using `yield` where we aren't anywhere else. Is there a functionality reason for it? If so, we should clarify that somewhere in the file, because I'm not 100% comfortable reviewing code with it since we haven't used it much before.

File: readthedocs/search/api/v3/backend.py [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+            project, version = self._get_project_and_version(value)
+
+            # Add the project itself.
+            if version and self._has_permission(self.request.user, version):
+                yield project, version
+
+            # If the user didn't provide a version, version_slug will be `None`,
+            # and we add all subprojects with their default version,
+            # otherwise we will add all projects that match the given version.
+            _, version_slug = self._split_project_and_version(value)
```
Comment (ericholscher) [inline readthedocs/search/api/v3/backend.py:102]: It's odd this logic is outside the `if project`. I think it should be inside?

File: readthedocs/search/api/v3/backend.py [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+                project=project,
+                version_slug=project.default_version,
+                include_hidden=False,
+            )
+            if version and self._has_permission(self.request.user, version):
+                yield project, version
+
+    def _get_subprojects(self, project, version_slug=None):
+        """
+        Get a tuple project/version of all subprojects of `project`.
```
Comment (ericholscher) [inline readthedocs/search/api/v3/backend.py:125]:
```suggestion
        Get a tuple (project,version) of all subprojects of `project`.
```

I thought it was returning a string as used in search, and was confused.

**Review by ericholscher — COMMENTED** [review]
This is a great refactor. I do think we need a bit more context on some of the decisions made. I commented in a few places to start, but I was also confused by the use of `yield`, which I'd like to understand better. If we doing new code approaches, we should be noting why it's necessary.

File: readthedocs/builds/querysets.py [hunk 19]
```diff
@@ -103,7 +103,9 @@ def public(
             if user.is_superuser:
                 queryset = self.all()
             else:
-                queryset = self._add_from_user_projects(queryset, user)
+                queryset = self._add_from_user_projects(
+                    queryset, user, admin=True, member=True
```
[the remark this answers is not in the document]
Comment (stsewd) [reply readthedocs/builds/querysets.py:107]: This matches what we have in .com, and actually in .com we override it completely (so this change doesn't affect .com),

https://github.com/readthedocs/readthedocs-corporate/blob/ad902765b50f570b44961ce8c8248e9fca713fb8/readthedocsinc/builds/querysets.py#L79-L84.

This was changed so the tests that I added with `RTD_ALLOW_ORGANIZATIONS=True` would pass on .org, otherwise this was returning an empty queryset (admin and member would be both false)

https://github.com/readthedocs/readthedocs.org/blob/adec99886dafd5fccba498bcab707b3a5cb0b40a/readthedocs/core/permissions.py#L47-L72

I thought I mentioned this in the PR description, but looks like I forgot, sorry

File: readthedocs/search/api/v3/urls.py (hunk 4 above)
Comment (stsewd) [reply readthedocs/search/api/v3/urls.py:6]: This is the same we are doing with the embed api https://github.com/readthedocs/readthedocs.org/blob/aa31543402a2c03e1f8a779e719266dc691b9671/readthedocs/embed/urls.py, and this also avoids having to repeat the `name="search_api_v3"` part when re-exporting it on .com.

File: readthedocs/search/api/v3/backend.py (hunk 6 above)
Comment (stsewd) [reply readthedocs/search/api/v3/backend.py:85]: it yields the results from `_get_default_projects`, after that it needs to stop, so a return is added.

File: readthedocs/search/api/v3/backend.py (hunk 7 above)
Comment (stsewd) [reply readthedocs/search/api/v3/backend.py:87]: This is the value of the `project` argument, it can be a project slug or a project slug + a version slug, e.g. `project` or `project/version`, wasn't sure about a better name for that.

File: readthedocs/search/api/v3/backend.py (hunk 8 above)
Comment (stsewd) [reply readthedocs/search/api/v3/backend.py:101]: if we don't a find a version with that slug, we fallback to the default version of that subproject

https://github.com/readthedocs/readthedocs.org/blob/adec99886dafd5fccba498bcab707b3a5cb0b40a/readthedocs/search/api/v3/backend.py#L123-L130

This is similar to what we do now when adding results from subprojects automatically

File: readthedocs/search/api/v3/views.py (hunk 9 above)
Comment (stsewd) [reply readthedocs/search/api/v3/views.py:59]: It felt more explicit since we are using it in several places inside this class

File: readthedocs/search/api/v3/views.py (hunk 10 above)
Comment (stsewd) [reply readthedocs/search/api/v3/views.py:104]: This isn't being overridden, it just a method of this class p: The confusion may be bc the `ListAPIView` has a list method, but this is just a `GenericView`, but we are using it to list search results, so I used a similar structure.

File: readthedocs/templates/core/project_bar_base.html (hunk 13 above)
Comment (stsewd) [reply readthedocs/templates/core/project_bar_base.html:44]: Would this be like a link on the detail view? I kind of like it having a direct way to search the project. Do we have a way to know how many people click/use this in GA? :thinking:

File: readthedocs/urls.py (hunk 14 above)
Comment (stsewd) [reply readthedocs/urls.py:94]: We have started using `path` on new code https://github.com/readthedocs/readthedocs.org/pull/9628/files.

And actually, the v2 URL was changed to `path` in another PR (that's the conflict the GitHub is showing on this PR)

File: readthedocs/search/api/v3/views.py (hunk 15 above)
Comment (stsewd) [reply readthedocs/search/api/v3/views.py:22]: Search doesn't require auth on .org, .com works with the auth backends, we could add support for `TokenAuthentication` as the other API v3 endpoints.

Not sure what you mean with semantics

File: readthedocs/search/api/v3/backend.py (hunk 16 above)
Comment (stsewd) [reply readthedocs/search/api/v3/backend.py:84]: We are using yield/iterators at

https://github.com/readthedocs/readthedocs.org/blob/aa31543402a2c03e1f8a779e719266dc691b9671/readthedocs/search/parsers.py#L134-L139

It's easier than adding things to a list and then returning it (we have several sources from where to add things), it also helps us to put a limit at the number of projects we can search at the same time https://github.com/readthedocs/readthedocs.org/blob/adec99886dafd5fccba498bcab707b3a5cb0b40a/readthedocs/search/api/v3/backend.py#L49

otherwise we would have the list of all projects and then cut it, here we just stop at the `n`th project.

File: readthedocs/templates/core/project_bar_base.html (hunk 13 above)
Comment (ericholscher) [reply readthedocs/templates/core/project_bar_base.html:44]: We should definitely have a nice way to search the project, but it feels odd as a UI element. We should find another way to expose it I think.

File: readthedocs/urls.py (hunk 14 above)
Comment (ericholscher) [reply readthedocs/urls.py:94]: It's just weird to have both of them in a single file. We should keep them all the same unless we do a bit migration of all  of them, otherwise it seems like there is a reason it's this way to readers.

File: readthedocs/search/api/v3/backend.py (hunk 16 above)
Comment (ericholscher) [reply readthedocs/search/api/v3/backend.py:84]: This is useful context, we should write it in the code somewhere.

File: readthedocs/templates/core/project_bar_base.html (hunk 13 above)
Comment (stsewd) [reply readthedocs/templates/core/project_bar_base.html:44]: @agjohnson any ideas/suggestions about where to put this `search this project` link?

File: readthedocs/search/api/v3/backend.py (hunk 16 above)
Comment (stsewd) [reply readthedocs/search/api/v3/backend.py:91]: Done!

File: readthedocs/search/api/v3/views.py (hunk 10 above)
Comment (ericholscher) [reply readthedocs/search/api/v3/views.py:107]: Good to note that then, also :)

**Comment by ericholscher** [comment]
@stsewd This feels close. You mentioned some width issues w/ the search box, do you have a link to a good screenshot of that?

File: readthedocs/templates/core/project_bar_base.html (hunk 13 above)
Comment (benjaoming) [reply readthedocs/templates/core/project_bar_base.html:44]: I was looking for a solution like this for the contextualized 404 pages :tada: 

So for now, I can refer to `{% url "elastic_project_search" project.slug %}` and then update it once the API is out?

**Comment by stsewd** [comment]
@ericholscher sorry, I thought I've attached screenshots in the PR description, looks like I didn't.

Screenshot 2022-11-07 at 16-49-18 Search project subproject project test-builds test Read the Docs
Screenshot 2022-11-07 at 16-48-48 Search project subproject test Read the Docs
Screenshot 2022-11-07 at 16-51-15 Search project subproject project test-builds search something more big Read the Docs

Basically, the text hides behind the "Search" button, and it also takes a lot of space that would be welcome when searching more than one project. We should probably just replace it with an icon and have the search bar not overlap with the button (not sure how this implemented in the new templates)

**Comment by ericholscher** [comment]
Yea, 👍 on either hiding the search box, or just moving it down to a new line.

File: readthedocs/search/api/v3/serializers.py [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+    with the following changes:
+
+    - ``project`` is an object, not a string.
+    - ``version`` is an object, not a string.
+    - ``project_alias`` isn't present,
+      it is contained in the ``project`` object.
+    """
+
+    project = serializers.SerializerMethodField()
+    version = serializers.SerializerMethodField()
```
Comment (humitos) [inline readthedocs/search/api/v3/serializers.py:23]: Why not creating a proper serializer for each of them instead of methods?

Example of this from APIv3 https://github.com/readthedocs/readthedocs.org/blob/8ccd55cf3370e772e8def502fe794386e887daf4/readthedocs/api/v3/serializers.py#L261

**Review by humitos — COMMENTED** [review]
I started doing a review but I noticed that it will take more time than I thought. It seems to be a big project. I'll come back to it with more time and focus.

File: readthedocs/templates/core/project_bar_base.html (hunk 13 above)
Comment (stsewd) [reply readthedocs/templates/core/project_bar_base.html:44]: yeah, that should be fine

File: readthedocs/search/api/v3/serializers.py (hunk 20 above)
Comment (stsewd) [reply readthedocs/search/api/v3/serializers.py:23]: The only one we can do this for is for the version, for the project we need to call `get_project_alias` to get the alias. The objects these serializers deal with are ES result objects, not RTD objects, we don't have that field in the ES object.

**Comment by stsewd** [comment]
@ericholscher I'm trying to fix the search bar, but looks like we are using it in other places as well (the same structure/css classes).


### Current state

Screenshot 2022-11-14 at 13-45-41 Versions Read the Docs
Screenshot 2022-11-14 at 13-45-27 Home Read the Docs
Screenshot from 2022-11-14 13-46-01

### My small fix

Looks like the problem is with the width being hardcoded (`width: 513px;`)

https://github.com/readthedocs/readthedocs.org/blob/5e870d46795f6f31da79fb11114bc9a41f1bba8b/media/css/core.css#L375-L375

So I tried to fix it setting the width to `width: 100%;`

Screenshot from 2022-11-14 13-41-02
Screenshot from 2022-11-14 13-41-26
Screenshot from 2022-11-14 13-41-48


The overlap is reduced, but now all forms have that small overlap.

Moving the button to a new line may require some more CSS...

Screenshot 2022-11-14 at 13-57-02 Search project test-builds Read the Docs

### New templates

The new templates won't/don't have this issue

Screenshot 2022-11-14 at 14-01-53 Search project test-builds - Read the Docs

If my small fix is enough, let me know. But I'm not sure if it's worth investing more time in that with the new templates coming :thinking:

**Comment by ericholscher** [comment]
@stsewd I don't think we need to invest much time in this. It seems fine as is, as well. It's not broken, so definitely don't invest much time here.

**Comment by ericholscher** [comment]
@stsewd I'm pretty happy with where this PR is. Is there any backwards incompatible changes required here, or can we roll it out and roll it back if needed?

**Comment by stsewd** [comment]
@ericholscher

- The doc structure was changed (redirects are needed in place, cc @benjaoming in case any of this collides with the work of moving our docs to diataxis)
- most of the facets are removed (dashboard search)
- Old project search links will redirect to the global search

One small detail that you may have an opinion on https://github.com/readthedocs/readthedocs.org/pull/9533#discussion_r1017804965.

What's next?

- Use api v3 in our embedded js search
- Use api v2 in our sphinx search extension

**Comment by benjaoming** [comment]
@stsewd we added this PR to the "iteration 0" milestone of the Diátaxis changes, so right now, I would hope that it's fixed up before Iteration 1 :+1:

**Comment by stsewd** [comment]
> so right now, I would hope that it's fixed up before Iteration 1

@benjaoming not sure what you mean with this p: should be fine to merge as is?

**Comment by benjaoming** [comment]
@stsewd 

> @benjaoming not sure what you mean with this p: should be fine to merge as is?

Yes absolutely, I wanted to clarify the below question

> cc @benjaoming in case any of this collides with the work of moving our docs to diataxis

The answer is basically no, please merge, and that even resolves one of the items in the Diátaxis milestone :+1: :tada:

**Comment by ericholscher** [comment]
> Use api v3 in our embedded js search

@stsewd What is this exactly? The indoc search that isn't using our extension?

**Comment by stsewd** [comment]
> I don't fully understand why our extension would use v2 instead of v3?

opps, that was my mistake, I meant to say "Use api v3 in our sphinx search extension"



>> Use api v3 in our embedded js search

> @stsewd What is this exactly? The indoc search that isn't using our extension?

yes

**Comment by benjaoming** [comment]
@stsewd I've added `Fixes #9573`

**Review by ericholscher — APPROVED** [review]
I think I'm OK merging this, and if I can get it tested on Monday that will be useful, but otherwise it feels safe to deploy. We can continue to iterate on it as we get real user feedback, but there's a bunch of improvements here that I'm excited about. 👍
